Docs: Fundamentals of Block Development - The block wrapper (#56596)
WordPress/gutenberg: docs/getting-started/fundamentals-block-development/the-block-wrapper.md

# The block wrapper

> Each block's markup is wrapped by a container HTML tag that needs to have the proper attributes to fully work in the Block Editor and to reflect the proper block's style settings when rendered in the Block Editor and the front end. As developers, when creating a custom block, we need to manually add these attributes to the markup using some of the tools provided by WordPress.

Reviewer: I find this sentence here a little confusing. It sounds like the markup of the block is wrapped by some markup we cannot control. 

Which was the case with the `apiVersion: 1` but with `apiVersion: 2 & 3` that no longer ist the case and we have the `useBlockProps` hook to get the attributes that need to exist on the wrapper to manually add to our own markup. So a developer has full control over the markup of the block including the wrapper

Author: @fabiankaegy I have rephrased that part inspired by the info in your comment.
Let me know what you think

Ensuring proper attributes to the block wrapper is especially important when using custom styling or features like `supports`. 

<div class="callout callout-info">
The use of <code>supports</code> generates a set of properties that need to be manually added to the wrapping element of the block so they're properly stored as part of the block data
</div>

> A block can have three markups defined, each one of them with a specific target and purpose:

Reviewer 2: ```suggestion
A block can have three sets of markup defined, each one of them with a specific target and purpose:
```

- The one for the **Block Editor**, defined through a `edit` React component passed to `registerBlockType` when registering the block in the client. 

> - The one used to **save the block in the DB**, defined through a `save` React component passed to `registerBlockType` when registering the block in the client. 

Reviewer 2: I've mentioned this in other review as well but `save` is not technically a React component but rather a function that return JSX. React components are [always uppercase](https://react.dev/learn/your-first-component#step-2-define-the-function). See [this conversation for context](https://github.com/WordPress/gutenberg/pull/38139)

Author: @ryanwelcher I have updated the texts to refer to `save` as a function rather than as a React component

    - This markup will be returned to the front end on request if no dynamic render has been defined for the block.
- The one used to **dynamically render the markup of the block** returned to the front end on request, defined through the `render_callback` on `register_block_type` or the `render` PHP file in `block.json`
    - If defined, this server-side generated markup will be returned to the front end, ignoring the markup stored in DB.

> For the React components `edit` and `save`, the block wrapper element should be a native DOM element (like `<div>`) or a React component that forwards any additional props to native DOM elements. Using a <Fragment> or <ServerSideRender> component, for instance, would be invalid.

Reviewer 2: ```suggestion
For the React component `edit` and the `save` function, the block wrapper element should be a native DOM element (like `<div>`) or a React component that forwards any additional props to native DOM elements. Using a <Fragment> or <ServerSideRender> component, for instance, would be invalid.
```

    data-type="block-development-examples/minimal-block-ca6eda" 
    data-title="Minimal Gutenberg Block ca6eda" 
    class="
        block-editor-block-list__block 
        wp-block 
        is-selected 
        wp-block-block-development-examples-minimal-block-ca6eda
    "
>Hello World - Block Editor</p>
```

Any additional classes and attributes for the `Edit` component of the block should be passed as an argument of `useBlockProps` (see [example](https://github.com/WordPress/block-development-examples/blob/trunk/plugins/stylesheets-79a4c3/src/edit.js)). When you add `support` for any feature, they get added to the object returned by the `useBlockProps` hook.

## The Save component's markup

When saving the markup in the DB, it’s important to add the block props returned by `useBlockProps.save()` to the wrapper element of your block. `useBlockProps.save()` ensures that the block class name is rendered properly in addition to any HTML attribute injected by the block supports API.

For example, for the following piece of code of a block's registration in the client that defines the markup desired for the DB (and returned to the front end by default)...

```js

> const Save = () => <p { ...useBlockProps.save() }>Hello World - Frontend</p>;

Reviewer 2: ```suggestion
const save = () => <p { ...useBlockProps.save() }>Hello World - Frontend</p>;
```
This will be confusing as all the [references use lower case](https://github.com/WordPress/block-development-examples/blob/trunk/plugins/stylesheets-79a4c3/src/save.js#L15)

registerBlockType( ..., {
	edit: Edit,

> 	save: Save,

Reviewer 2: ```suggestion
	save,
```

} );
```

_(see the [code above](https://github.com/WordPress/block-development-examples/blob/trunk/plugins/minimal-block-ca6eda/src/index.js) in [an example](https://github.com/WordPress/block-development-examples/tree/trunk/plugins/minimal-block-ca6eda))_

...the markup of the block in the front end could look like this:
```html
<p class="wp-block-block-development-examples-minimal-block-ca6eda">Hello World – Frontend</p>
```

> Any additional classes and attributes for the `Save` component of the block should be passed as an argument of `useBlockProps.save()` (see [example](https://github.com/WordPress/block-development-examples/blob/trunk/plugins/stylesheets-79a4c3/src/save.js)). 

Reviewer 2: ```suggestion
Any additional classes and attributes for the `save` function of the block should be passed as an argument of `useBlockProps.save()` (see [example](https://github.com/WordPress/block-development-examples/blob/trunk/plugins/stylesheets-79a4c3/src/save.js)). 
```
